corehq/apps/reports/tests/test_case_data.py

```diff
@@ -0,0 +1,94 @@
+        self._create_case_property("prop2", "group1")
+        result = _get_dd_tables(self.domain, self.case_type, {}, self.timezone)
+        self._assert_grouping_order_tables(result, [{"name": "group1", "properties": ["prop1", "prop2"]}])
+
+    def test_dd_ordering_when_multiple_groups(self):
+        """
+        Tests that when a data dictionary contains multiple user defined groups,
+        ordering of groups and properties within the group is retained.
+        """
+        # Create group1 properties
+        self._create_case_property("prop1", "group1")
+        self._create_case_property("prop2", "group1")
+        self._create_case_property("prop3", "group2")
+        self._create_case_property("prop4", "group2")
+        result = _get_dd_tables(self.domain, self.case_type, {}, self.timezone)
+        self._assert_grouping_order_tables(result, [{"name": "group1", "properties": ["prop1", "prop2"]},
+                                                    {"name": "group2", "properties": ["prop3", "prop4"]}])
+
+    def test_dd_ordering_when_multiple_groups_and_unrecognized_group(self):
+        """
+        Tests that when a data dictionary contains multiple user defined groups and a unrecognized group,
+        ordering of groups and properties within the group is retained. Ordering is not retained in
+        unrecognized group.
+        """
+        # Create group1 properties
+        self._create_case_property("prop1", "group1")
+        self._create_case_property("prop2", "group1")
+        self._create_case_property("prop3", "group2")
+        self._create_case_property("prop4", "group2")
+
+        # Dynamic data with additional property not present in any of the groups
+        dynamic_data = {"prop1": None, "prop2": None, "prop3": None, "prop4": None, "prop5": None, "prop6": None}
+
+        result = _get_dd_tables(self.domain, self.case_type, dynamic_data, self.timezone)
+        self._assert_grouping_order_tables(result, [{"name": "group1", "properties": ["prop1", "prop2"]},
+                                                    {"name": "group2", "properties": ["prop3", "prop4"]},
+                                                    {"name": "Unrecognized", "properties": ["prop5", "prop6"]}
+                                                    ])
+
+    def _assert_grouping_order_tables(self, result, expected_result):
+        """
+        Sample result
+        [{'name': 'group1', 'rows': [[{'expr': 'prop1', 'name': 'prop1', 'description': '', 'value': '---',
+        'has_history': True}, {'expr': 'prop2', 'name': 'prop2', 'description': '', 'value': '---',
+        'has_history': True}]]}, {'name': 'group2', 'rows': [[{'expr': 'prop3', 'name': 'prop3',
+        'description': '', 'value': '---', 'has_history': True}, {'expr': 'prop4', 'name': 'prop4',
+        'description': '', 'value': '---', 'has_history': True}]]}]
+        """
+        idx = 0
+        for group_data in result:
```

Comment:
nit: you can zip these two things together so you don't need to worry about indices
```suggestion
        self.assertEqual(len(result), len(expected_result))
        for group_actual, group_expected in zip(result, expected_result):
```